### ISSUE TITLE
Moving the post subject to the topics table.

### ISSUE THREAD

**Issue by Vekseid** [elkarte/Elkarte#2024]
So, I had a little rant over on the Elkarte forums about the heavy denormalization in the threads table.

This frequently creates scenarios in which the first and last messages  and/or the post count gets out of sync in the topics table, which causes user posts to vanish, and if you have a forum of any meaningful size, you have to write your own fix script for this, because running the whole error checking script simply does not work.

My proposal is to do the following
1) Move the subject column into the topics table
2) Remove all message ids from the topics table
3) Put the needed denormalized data into the topics table in their place (last poster, last post/edit time, etc.)

This won't fix everything, but the separate subject-per-post annoys most people, and this will help keep programmers from attempting to do 'optimizations' like trying to bound by first and last post id rather than rely on the thread index (ahem).

**Comment by Spuds** [comment]
> but the separate subject-per-post annoys most people

I never quite understood why this was allowed, not sure if it was an SMF only feature or done that way in other forum packages, or just done cuz ya could do to the way the db is structured.

> Remove all message ids from the topics table

So you are proposing that id_first_message and id_last_message in the topics table not be tracked/logged but be found/calculated as needed?

**Comment by Vekseid** [comment]
The proposal is that the stuff we need from them for the topics index (First and last poster, topic start, last post, and last edit times) be stored in the topics table, so that the messages table only needs to get loaded when the thread is viewed. This is generally ephemeral data and it's not too big a deal for it to get out of sync every now and then (compared to what happens when the last message id or post count gets out of line).

**Comment by joshuaadickerson** [comment]
This has been discussed at SMF and before that YaBB SE and before that YaBB for over a decade. It has always been knocked down because people would lose context in some posts. I think it make total sense and the only thing that needs to be figured out is what to do with people's posts where they used a different subject. I think adding either a subject tag and either prepending or appending it to the post, a mod, having the admin go through each post that has a different subject, or just letting everyone know that this is a backwards incompatible change.

**Comment by Targren** [comment]
Unfortunately, I'm sitting on the cold floor of a colo at the moment and can't look at the database schema I have set up on my dev box back home, so I'm working on context of what I've read here: It sounds like the message record has a thread/topic ID field, or is there a standalone table that associates each message ID with at topic ID? 

Please tell me the topics don't just store first and last, and the rest of the message use "parent message IDs" like email clients...

**Comment by emanuele45** [comment]
I suggested something similar a while ago (and I thought there was a discussion on the forum, but I can't find it), it's true that in some cases is useful to have a different subject for each post (see for example [this](http://www.elkarte.net/community/index.php?topic=831.0) or [this](http://www.elkarte.net/community/index.php?topic=1957.0) topics), but on the other hand the function is used so seldom that is probably not worth the hassles...

Just a minor remark:
>  so that the messages table only needs to get loaded when the thread is viewed.

This is probably unlikely to happen unless message previews are removed as "inline feature" and are moved to an AJAX call. Because the body of the first or last message has to be taken from the messages table anyway.

**Comment by Targren** [comment]
>  >  so that the messages table only needs to get loaded when the thread is viewed.

> This is probably unlikely to happen unless message previews are removed as "inline feature" and are moved to an AJAX call. Because the body of the first or last message has to be taken from the messages table anyway.

It's probably not worth the effort, anyway. Unless they're using SQLite on the backend, even MySQL does a pretty good job of simple joins these days, unless you really mangle the query.

**Comment by emanuele45** [comment]
> Unfortunately, I'm sitting on the cold floor of a colo at the moment and can't look at the database schema

http://www.elkarte.it/db/schema.php

:laughing:

**Comment by Targren** [comment]
Yeah, as I expected, there's a lot of stuff in the topics table that's duplicated data. :P

**Comment by joshuaadickerson** [comment]
Denormalization isn't bad. Often it is good.

**Comment by Targren** [comment]
It can be good, or at least appropriate, when used correctly, yes. 

The fact that the tables tend to go out of sync (which is what led to this request) suggest it might not be getting used correctly, though.

**Comment by joshuaadickerson** [comment]
As far as I can remember, I've never heard of a topic getting out of sync. Are there any examples? If the last_post_id is getting out of sync, I can think of a simple way to fix that. Just check to make sure that the last_post_id < $new_last_post_id or whatever way you want to put that condition.

**Comment by Targren** [comment]
Take a look at Vekseid's original post describing the issues he's seen.

**Comment by Vekseid** [comment]
99% of the time the different subject is something off the wall and completely ignored until someone notices it a hundred or thousand posts later.

> I suggested something similar a while ago (and I thought there was a discussion on the forum, but I can't find it), it's true that in some cases is useful to have a different subject for each post (see for example this or this topics),

I made an anchor tag for Elliquiy to help members organize not only by post but within posts as well. And it's not like this would prevent linking specific replies.

> As far as I can remember, I've never heard of a topic getting out of sync. Are there any examples? If the last_post_id is getting out of sync, I can think of a simple way to fix that. Just check to make sure that the last_post_id < $new_last_post_id or whatever way you want to put that condition.

I'm not sure why you think that would fix a last post id getting clobbered.

It's happened with some regularity on E, along with post count problems.
